receiver/snmpreceiver/scraper.go

```diff
@@ -0,0 +1,559 @@
+				if !strings.HasPrefix(oid.OID, ".") {
+					oid.OID = "." + oid.OID
+					s.cfg.Metrics[name].ScalarOIDs[i].OID = oid.OID
+				}
+				scalarMetricOIDs = append(scalarMetricOIDs, oid.OID)
+				scalarMetricNamesByOID[oid.OID] = name
+			}
+		}
+	}
+
+	// If no scalar metric configs, nothing else to do
+	if len(scalarMetricOIDs) == 0 {
+		return nil
+	}
+
+	// Get all SNMP scalar OID data and turn it into metrics/attributes
+	// which are then stored in the passed in metricsMap
+	err = s.client.GetScalarData(scalarMetricOIDs, scalarDataToMetric(now, metricsMap, scalarMetricNamesByOID, s.cfg))
+
+	return err
+}
+
+// scalarDataToMetric provides a function which will convert one piece of SNMP scalar data, turn it into
+// a metric with attributes based on the related configs, store it in the passed in metricsMap
+func scalarDataToMetric(
+	now pcommon.Timestamp,
+	metricsMap map[string]*pmetric.Metric,
+	scalarMetricNamesByOID map[string]string,
+	cfg *Config,
+) processFunc {
+	// This returns a processFunc because this is what the client's GetScalarData method requires
+	return func(data snmpData) error {
+		// Return an error if this SNMP scalar data is not of a useable type
+		switch data.valueType {
+		case NotSupported:
+			fallthrough
+		case String:
+			return fmt.Errorf(errMsgBadValueType, data.oid)
+		}
+
+		// Retrieve the metric config for this SNMP data
+		metricName := scalarMetricNamesByOID[data.oid]
+		metricCfg := cfg.Metrics[metricName]
+
+		// Get all enum attributes names and values for this SNMP scalar data based on the metric config
+		var metricAttributes []Attribute
+		for _, scalarOID := range metricCfg.ScalarOIDs {
+			if scalarOID.OID == data.oid {
+				metricAttributes = scalarOID.Attributes
+			}
+		}
+
+		// Get/create the metric and datapoint for this SNMP scalar data and make sure metricsMap is current
+		metric := metricsMap[metricName]
+		metric, dp := createNewMetricDataPoint(now, data, metric, metricName, metricCfg)
+		if metricsMap[metricName] == nil {
+			metricsMap[metricName] = metric
+		}
+
+		// Set enum attributes for this metric's datapoint based on the previously gathered attributes.
+		// Keys will be determined from the related attribute config and enum values will come straight from
+		// the metric config's attributes.
+		for _, attribute := range metricAttributes {
+			attributeCfg := cfg.Attributes[attribute.Name]
+			attributeKey := attribute.Name
+			if attributeCfg.Value != "" {
+				attributeKey = attributeCfg.Value
+			}
+			dp.Attributes().PutString(attributeKey, attribute.Value)
+		}
+
+		return nil
+	}
+}
+
+// scrapeIndexedMetrics retrieves all SNMP data from column OIDs and turns the returned indexed data
+// into metrics with optional attribute and/or resource attributes
+func (s *snmpScraper) scrapeIndexedMetrics(
+	now pcommon.Timestamp,
+	resourceMetricsMap map[string]*pmetric.ResourceMetrics,
+	metricsByResourceMap map[string]map[string]*pmetric.Metric,
+) (err error) {
```

Comment:
nit: I would remove the named return there doesn't seem to be any place we're using it as a return value.

Reply:
changed